gpytorch/kernels/arc_kernel.py

```diff
@@ -0,0 +1,185 @@
+                              angle_prior=gpytorch.priors.GammaPrior(0.5,1),
+                              radius_prior=gpytorch.priors.GammaPrior(3,2),
+                              ard_num_dims=x.shape[-1])),
+                num_dims=x.shape[-1])
+        >>> covar = covar_module(x
+        >>> print(covar.shape)
+    """
+
+    has_lengthscale = True
+
+    def __init__(
+        self, base_kernel, angle_prior: Optional[Prior] = None, radius_prior: Optional[Prior] = None, **kwargs
+    ):
+        super().__init__(has_lengthscale=True, **kwargs)
+
+        # TODO: check the errors given by interval
+        angle_constraint = Positive()
+
+        self.register_parameter(
+            name="raw_angle", parameter=torch.nn.Parameter(torch.zeros(*self.batch_shape, 1, self.ard_num_dims)),
+        )
+        if angle_prior is not None:
+            self.register_prior(
+                "angle_prior", angle_prior, lambda: self.angle, lambda v: self._set_angle(v),
+            )
+
+        self.register_constraint("raw_angle", angle_constraint)
+
+        self.register_parameter(
+            name="raw_radius", parameter=torch.nn.Parameter(torch.zeros(*self.batch_shape, 1, self.ard_num_dims)),
+        )
+
+        if radius_prior is not None:
+            self.register_prior(
+                "radius_prior", radius_prior, lambda: self.radius, lambda v: self._set_radius(v),
+            )
+
+        radius_constraint = Positive()
+        self.register_constraint("raw_radius", radius_constraint)
+
+        self.base_kernel = base_kernel
+        if self.base_kernel.has_lengthscale:
+            self.base_kernel.lengthscale = 1
+            self.base_kernel.raw_lengthscale.requires_grad_(False)
+
+    @property
+    def angle(self):
+        return self.raw_angle_constraint.transform(self.raw_angle)
+
+    @angle.setter
+    def angle(self, value):
+        self._set_angle(value)
+
+    def _set_angle(self, value):
+        if not torch.is_tensor(value):
+            value = torch.as_tensor(value).to(self.raw_angle)
+        self.initialize(raw_angle=self.raw_angle_constraint.inverse_transform(value))
+
+    @property
+    def radius(self):
+        return self.raw_radius_constraint.transform(self.raw_radius)
+
+    @radius.setter
+    def radius(self, value):
+        self._set_radius(value)
+
+    def _set_radius(self, value):
+        if not torch.is_tensor(value):
+            value = torch.as_tensor(value).to(self.raw_radius)
+        self.initialize(raw_radius=self.raw_radius_constraint.inverse_transform(value))
+
+    def embedding(self, x):
+        x_ = x.div(self.lengthscale)
+        x_s = self.radius * torch.sin(pi * self.angle * x_)
+        x_c = self.radius * torch.cos(pi * self.angle * x_)
+        x_ = torch.cat((x_s, x_c), dim=-1).squeeze(0)
+        return x_
+
+    def forward(self, x1, x2, diag=False, **params):
+        x1_, x2_ = self.embedding(x1), self.embedding(x2)
+        return self.base_kernel(x1_, x2_)
```

Comment:
You might need to pass diag onwards in the base kernel. Something like:
```suggestion
        return self.base_kernel(x1_, x2_, diag=diag)
```
You also may need to just verify batch support. If you'd like, you can extend our base kernel test case here:
https://github.com/cornellius-gp/gpytorch/blob/25e9f99f737eb8967580eda6e8c338b2311e13a6/gpytorch/test/base_kernel_test_case.py#L8

If you fill in the two abstract methods, for example like here: https://github.com/cornellius-gp/gpytorch/blob/25e9f99f737eb8967580eda6e8c338b2311e13a6/test/kernels/test_rbf_kernel.py#L12-L17
then the base kernel test case tests generally all the settings your kernel might get called with across all sorts of GPyTorch models. If you pass all of those tests, then you would know the kernel is in pretty good shape.

Reply:
I am going to try to build the test cases. Asa I can I will upload them. Thank you very much :)

Reply:
Hi, the kernel now passes the tests. There were problems with adding `self.ard_num_dims` as last dimension for the tensor registerer of radius and angle. I have put a conditional to solve this, but in theory the arc kernel should be used to always with `ard_num_dims`; at least is what makes sense to me. By the way, I have also included the test, but then the checks fail because it cannot find the arc kernel. Should I add also the test or should I add only the kernel?

Reply:
Hi @BCJuan,

You need to add something like `from .arc_kernel import ArcKernel` to `gpytorch/kernels/__init__.py`. This will enable `from gpytorch.kernels import ArcKernel` to work, and will also make your test pass.

Reply:
Aps, still wondering how could not see that. My apologies and thank you for the clarification. Fixed and added.

Reply:
I am thinking that to further check the functionality of the arc kernel I could do something of what is found in `test_rbf_kernel.py` in functions such as `test_ard` or `test_ard_batch`, meaning checking an actual computation with the kernel implementation. As in `rest_rbf_kernel` they would go in the test for the arc kernel.